caluma/caluma_form/models.py

```diff
@@ -422,33 +422,86 @@ def copy(self, family=None, user=None):
+        # method.
+        if is_main_doc:
+            # get all dynamic answers of the document family
+            family_dynamic_answers = Answer.objects.filter(
+                document__in=Document.objects.filter(family=family).values_list(
+                    "pk", flat=True
+                ),
+                question__type__in=[
+                    Question.TYPE_DYNAMIC_CHOICE,
+                    Question.TYPE_DYNAMIC_MULTIPLE_CHOICE,
+                ],
+            )
+
+            for new_answer in family_dynamic_answers:
+                answer_value_changed = False
+
+                # initialize the data source used by the dynamic question
+                data_source_class = get_data_sources(dic=True)[
+                    new_answer.question.data_source
+                ]
+                data_source = data_source_class()
+
+                for new_dynamic_option in DynamicOption.objects.filter(
+                    document=new_answer.document, question=new_answer.question
+                ):
+                    # get the old answer from the source document for comparison
+                    old_answer = Answer.objects.get(
+                        document=new_answer.document.source,
+                        question=new_answer.question,
+                    )
+
+                    # let the data source decide what to do with the answer value.
+                    new_slug, new_label = data_source.on_copy(
+                        old_answer=old_answer,
+                        new_answer=new_answer,
+                        old_value=(new_dynamic_option.slug, new_dynamic_option.label),
+                    )
+
+                    # modify the answer value if the datasource decides to change or
+                    # discard the answer value.
+                    if new_slug is None or new_slug != new_dynamic_option.slug:
+                        answer_value_changed = True
+                        new_answer.value = new_answer.modify_changed_choice_answer(
+                            new_answer.question,
+                            new_dynamic_option.slug,
+                            new_slug,
+                            new_answer.value,
+                        )
+
+                    # update the dynamic option if the slug has changed
+                    if new_slug and new_slug != new_dynamic_option.slug:
+                        new_dynamic_option.slug = new_slug
+                        new_dynamic_option.label = new_label
+                        new_dynamic_option.save()
+
+                    # delete the dynamic option if the slug is None
+                    elif not new_slug:
+                        new_dynamic_option.delete()
+
+                # save the answer value if changed after checking all dynamic options
+                if answer_value_changed:
+                    new_answer.save(update_fields=["value"])
+
```

Comment:
I think this whole block of code should be moved into it's own method - the method is otherwise way too overloaded. Create a new method `_update_dynamic_answers_after_copy()` (or similar - maybe you have a better idea for a name :-) )